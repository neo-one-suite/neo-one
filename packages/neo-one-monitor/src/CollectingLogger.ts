import { CollectedLoggerLogOptions, Logger, LoggerLogOptions } from './types';

export class CollectingLogger implements Logger {
  private logs: CollectedLoggerLogOptions[] = [];

  public log(options: LoggerLogOptions): void {
    const logOptions = {
      name: options.name,
      level: options.level,
      message: options.message,
      labels: options.labels,
      data: options.data,
      error:
        options.error == null
          ? undefined
          : {
              message: options.error.message,
              stack: options.error.stack,
              code:
                (options.error as any).code == null
                  ? options.error.constructor.name
                  : (options.error as any).code,
            },
    };

    if (logOptions.error == null) {
      delete logOptions.error;
    }

    this.logs.push(logOptions);
  }

  public collect(): CollectedLoggerLogOptions[] {
    const { logs } = this;
    this.logs = [];

    return logs;
  }

  public close(callback: () => void): void {
    callback();
  }
}
