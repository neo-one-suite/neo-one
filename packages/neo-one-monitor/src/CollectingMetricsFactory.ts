import {
  CollectedMetric,
  Counter,
  Histogram,
  Labels,
  MetricCollection,
  MetricOptions,
  BucketedMetricOptions,
  PercentiledMetricOptions,
  MetricValue,
  CounterBase,
  HistogramBase,
} from './types';
import {
  CounterProxy,
  HistogramProxy,
  MetricsFactoryProxy,
} from './MetricsFactoryProxy';

class CollectingMetric<T extends MetricOptions | BucketedMetricOptions> {
  public readonly metric: T;
  protected values: MetricValue[];

  constructor(metric: T) {
    this.metric = metric;
    this.values = [];
  }

  public toJSON(): CollectedMetric<T> {
    return { metric: this.metric, values: this.values };
  }

  public reset(): void {
    this.values = [];
  }
}

class CollectingCounter extends CollectingMetric<MetricOptions>
  implements CounterBase {
  public inc(countOrLabels?: Labels | number, count?: number): void {
    this.values.push({ countOrLabels, count });
  }
}

class CollectingHistogram extends CollectingMetric<BucketedMetricOptions>
  implements HistogramBase {
  public observe(countOrLabels?: Labels | number, count?: number): void {
    this.values.push({ countOrLabels, count });
  }
}

export class CollectingMetricsFactory extends MetricsFactoryProxy {
  private counters: { [name: string]: CollectingCounter } = {};
  private histograms: { [name: string]: CollectingHistogram } = {};

  public collect(): MetricCollection {
    const currentMetrics = this.serializeJSON();
    Object.values(this.counters).forEach((metric) => metric.reset());
    Object.values(this.histograms).forEach((metric) => metric.reset());

    return currentMetrics;
  }

  public reset_forTest(): void {
    this.counters = {};
    this.histograms = {};
  }

  protected createCounterInternal(options: MetricOptions): Counter {
    const counter = new CollectingCounter(options);
    this.counters[counter.metric.name] = counter;

    return this.initializeMetric(
      new CounterProxy(counter, options.labelNames),
      options,
      (metric) => metric.inc(0),
      (metric, labels) => metric.inc(labels, 0),
    );
  }

  protected createHistogramInternal(options: BucketedMetricOptions): Histogram {
    const histogram = new CollectingHistogram(options);
    this.histograms[histogram.metric.name] = histogram;

    return this.initializeMetric(
      new HistogramProxy(histogram, options.labelNames),
      options,
      (metric) => metric.observe(0),
      (metric, labels) => metric.observe(labels, 0),
    );
  }

  private serializeJSON(): MetricCollection {
    return {
      counters: Object.entries(this.counters).reduce(
        (accMetric, [name, metric]) => ({
          ...accMetric,
          [name]: metric.toJSON(),
        }),
        {},
      ),
      histograms: Object.entries(this.histograms).reduce(
        (accMetric, [name, metric]) => ({
          ...accMetric,
          [name]: metric.toJSON(),
        }),
        {},
      ),
    };
  }

  private initializeMetric<T extends CounterBase | HistogramBase>(
    metric: T,
    {
      labelNames,
      labels = [],
    }: MetricOptions | BucketedMetricOptions | PercentiledMetricOptions,
    init: (metric: T) => void,
    initLabels: (metric: T, labels: Labels) => void,
  ): T {
    if (labelNames == null) {
      init(metric);
    } else {
      for (const labelSet of labels) {
        initLabels(metric, labelSet);
      }
    }

    return metric;
  }
}

export const collectingMetrics: CollectingMetricsFactory = new CollectingMetricsFactory();
