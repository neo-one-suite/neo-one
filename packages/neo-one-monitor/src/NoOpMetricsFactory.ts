import { MetricsFactoryProxy } from './MetricsFactoryProxy';
import { MetricsFactory } from './types';

class NoOpMetricsFactory extends MetricsFactoryProxy {}

export const metrics: MetricsFactory = new NoOpMetricsFactory();
