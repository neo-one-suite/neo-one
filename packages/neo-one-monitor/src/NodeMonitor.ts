import { IncomingMessage } from 'http';
import Koa, { Context } from 'koa';
import { Server } from 'net';

import gcStats from 'prometheus-gc-stats';
import mount from 'koa-mount';
import prom from 'prom-client';

import { Logger, LogLevel, Monitor } from './types';
import { MonitorBase, Tracer } from './MonitorBase';

export interface NodeMonitorCreate {
  service: string;
  logger?: Logger;
  tracer?: Tracer;
  spanLogLevel?: LogLevel;
}

export class NodeMonitor extends MonitorBase {
  public static create({
    service,
    logger,
    tracer,
    spanLogLevel,
  }: NodeMonitorCreate): NodeMonitor {
    prom.collectDefaultMetrics({ timeout: 4000 });
    gcStats(prom.register)();
    return new NodeMonitor({
      service,
      component: service,
      logger: logger || {
        log: () => {
          // do nothing
        },
        close: (callback: () => void) => {
          callback();
        },
      },
      tracer,
      // Perfhooks is broken in 8.9 - there's no way to get the current
      // high resolution timestamp. Fix is in 9.8.0.
      now: () => Date.now(),
      spanLogLevel,
    });
  }

  private server: Server | null = null;

  public forContext(ctx: Context): Monitor {
    return this.withLabels({
      [this.labels.HTTP_METHOD]: ctx.request.method,
      [this.labels.SPAN_KIND]: 'server',
      [this.labels.HTTP_REQUEST_PROTOCOL]: ctx.request.protocol,
    }).withData({
      [this.labels.HTTP_HEADERS]: JSON.stringify(ctx.request.headers),
      [this.labels.HTTP_URL]: ctx.request.originalUrl,
      [this.labels.HTTP_FULLPATH]: ctx.request.path,
      [this.labels.HTTP_REQUEST_QUERY]: ctx.request.querystring,
      [this.labels.PEER_ADDRESS]: ctx.request.ip,
      [this.labels.PEER_PORT]: ctx.request.socket.remotePort,
      [this.labels.HTTP_REQUEST_SIZE]: ctx.request.length,
    });
  }

  public forMessage(message: IncomingMessage): Monitor {
    const app = new Koa();
    app.proxy = true;
    // $FlowFixMe
    app.silent = true;
    const ctx = app.createContext(message, undefined as any);
    return this.forContext(ctx);
  }

  public serveMetrics(port: number): void {
    const app = new Koa();
    app.proxy = true;
    // $FlowFixMe
    app.silent = true;

    const monitor = this.at('telemetry');
    app.on('error', (error) => {
      monitor.logError({
        name: 'http_server_request_uncaught_error',
        message: 'Unexpected uncaught request error.',
        error,
      });
    });

    app.use(
      mount('/metrics', (ctx: Context) => {
        ctx.body = prom.register.metrics();
      }),
    );

    this.server = app.listen(port);
  }

  protected async closeInternal(): Promise<void> {
    clearInterval(prom.collectDefaultMetrics());
    await Promise.all([
      super.closeInternal(),
      new Promise((resolve) => {
        if (this.server == null) {
          resolve();
        } else {
          this.server.close(() => resolve());
        }
      }),
    ]);
  }
}
