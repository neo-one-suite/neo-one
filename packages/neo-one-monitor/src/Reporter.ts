import { Subscription, interval } from 'rxjs';

import { mergeScan } from 'rxjs/operators';

import { CollectingLogger } from './CollectingLogger';
import { Counter, Report } from './types';

import { collectingMetrics } from './CollectingMetricsFactory';
import { metrics as metricsFactory } from './NoOpMetricsFactory';

const MAX_BACKLOG = 1000;

const EMPTY_BACKREPORT: Report = {
  logs: [],
  metrics: {
    counters: {},
    histograms: {},
  },
};

export interface CollectReportOptions {
  backReport?: Report;
  logger: CollectingLogger;
}

export class Reporter {
  private readonly endpoint: string;
  private readonly subscription: Subscription;
  private readonly requestsTotal: Counter;
  private readonly requestErrorsTotal: Counter;

  constructor({
    logger,
    timer,
    endpoint,
  }: {
    logger: CollectingLogger;
    timer: number;
    endpoint: string;
  }) {
    this.endpoint = endpoint;
    this.requestsTotal = metricsFactory.createCounter({
      name: 'http_report_metrics_requests_total',
    });
    this.requestErrorsTotal = metricsFactory.createCounter({
      name: 'http_report_metrics_failures_total',
    });

    this.subscription = interval(timer)
      .pipe(
        mergeScan(
          async (backReport) =>
            this.collectReport({
              backReport,
              logger,
            }),
          EMPTY_BACKREPORT,
          1,
        ),
      )
      .subscribe();
  }

  public close(): void {
    this.subscription.unsubscribe();
  }

  public async collectReport_forTest(
    options: CollectReportOptions,
  ): Promise<Report> {
    return this.collectReport(options);
  }

  private async report(report: Report): Promise<Response> {
    return fetch(this.endpoint, {
      method: 'POST',
      body: JSON.stringify(report),
    });
  }

  private async collectReport({
    backReport,
    logger,
  }: CollectReportOptions): Promise<Report> {
    const logs = logger.collect();
    const metrics = collectingMetrics.collect();

    const report = this.addBackReport({ logs, metrics }, backReport);

    let response;
    try {
      response = await this.report(report);
    } catch (error) {
      response = { ok: false };
    }

    this.requestsTotal.inc();
    if (!response.ok) {
      this.requestErrorsTotal.inc();
      return this.constructBackReport(report);
    }

    return EMPTY_BACKREPORT;
  }

  private addBackReport(report: Report, backReport?: Report): Report {
    if (backReport == null) {
      return report;
    }

    return {
      logs: backReport.logs.concat(report.logs),
      metrics: {
        counters: Object.entries(backReport.metrics.counters).reduce(
          (accMetric, [name, metric]) => ({
            ...accMetric,
            [name]: {
              metric: metric.metric,
              values:
                accMetric[name] == null
                  ? metric.values
                  : metric.values.concat(accMetric[name].values),
            },
          }),
          report.metrics.counters,
        ),
        histograms: Object.entries(backReport.metrics.histograms).reduce(
          (accMetric, [name, metric]) => ({
            ...accMetric,
            [name]: {
              metric: metric.metric,
              values:
                accMetric[name] == null
                  ? metric.values
                  : metric.values.concat(accMetric[name].values),
            },
          }),
          report.metrics.histograms,
        ),
      },
    };
  }

  private constructBackReport(report: Report): Report {
    return {
      logs: report.logs.slice(-MAX_BACKLOG),
      metrics: {
        counters: Object.entries(report.metrics.counters).reduce(
          (accMetric, [name, metric]) => ({
            ...accMetric,
            [name]: {
              metric: metric.metric,
              values: metric.values.slice(-MAX_BACKLOG),
            },
          }),
          {},
        ),
        histograms: Object.entries(report.metrics.histograms).reduce(
          (accMetric, [name, metric]) => ({
            ...accMetric,
            [name]: {
              metric: metric.metric,
              values: metric.values.slice(-MAX_BACKLOG),
            },
          }),
          {},
        ),
      },
    };
  }
}
