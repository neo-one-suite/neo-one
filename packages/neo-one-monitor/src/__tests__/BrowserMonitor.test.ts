/* @jest-environment jsdom */
import prom from 'prom-client';

import { CollectingLogger } from '../CollectingLogger';
import { NodeMonitor } from '../NodeMonitor';
import { Reporter } from '../Reporter';

import { collectingMetrics } from '../CollectingMetricsFactory';
import { metrics } from '../NoOpMetricsFactory';
import { LoggerLogOptions } from '../types';

describe('BrowserMonitor', () => {
  const options1: LoggerLogOptions = {
    name: 'log1',
    level: 'info',
    message: 'test1',
    labels: { label1: 10 },
    data: { data1: '20' },
  };
  const error = new Error('testError');
  (error as any).code = 'red';
  const options2: LoggerLogOptions = {
    name: 'log2',
    level: 'silly',
    message: 'test2',
    labels: { label2: 30 },
    data: { data1: '40' },
    error,
  };

  const options2Result = {
    ...options2,
    error: {
      message: error.message,
      stack: error.stack,
      code: (error as any).code,
    },
  };

  const counterOptions1 = {
    name: 'counter1',
    help: 'helpCount',
    labelNames: ['label1', 'label2'],
  };

  const counterResult1 = {
    metric: {
      name: counterOptions1.name,
      help: counterOptions1.help,
      labelNames: [...counterOptions1.labelNames],
    },
    values: [{ countOrLabels: 5 }, { countOrLabels: 4 }],
  };

  const histogramOptions1 = {
    name: 'histogram1',
    help: 'helpHist',
    labelNames: ['histLabel1', 'histLabel2'],
  };

  const histResult1 = {
    metric: {
      name: histogramOptions1.name,
      help: histogramOptions1.help,
      labelNames: [...histogramOptions1.labelNames],
    },
    values: [{ countOrLabels: { histLabel1: 14 }, count: 3 }],
  };

  const emptyMetrics = {
    counters: {
      [counterResult1.metric.name]: {
        metric: counterResult1.metric,
        values: [],
      },
    },
    histograms: {
      [histResult1.metric.name]: {
        metric: histResult1.metric,
        values: [],
      },
    },
  };

  const requestCounter = {
    name: 'http_report_metrics_requests_total',
  };

  const failuresCounter = {
    name: 'http_report_metrics_failures_total',
  };

  const emptyBackReport = {
    logs: [],
    metrics: {
      counters: {},
      histograms: {},
    },
  };

  let logger = new CollectingLogger();
  let reporter: Reporter | null;
  beforeEach(() => {
    logger = new CollectingLogger();
    reporter = null;
    metrics.setFactory(collectingMetrics);
  });

  afterEach(() => {
    collectingMetrics.reset_forTest();
    prom.register.clear();
    if (reporter != null) {
      reporter.close();
    }
  });

  test('CollectingLogger', () => {
    expect(logger.collect()).toEqual([]);

    logger.log(options1);
    const firstResult = logger.collect();
    expect(firstResult).toEqual([options1]);

    logger.log(options2);
    const secondResult = logger.collect();
    expect(secondResult).toEqual([options2Result]);

    expect(logger.collect()).toEqual([]);

    const closeFunc = jest.fn();

    logger.close(closeFunc);

    expect(closeFunc).toHaveBeenCalledTimes(1);
  });

  test('BrowserMetricsFactory', () => {
    const counter = metrics.createCounter(counterOptions1);
    counter.inc(5);
    counter.inc(4);

    const histogram = metrics.createHistogram(histogramOptions1);
    histogram.observe({ histLabel1: 14 }, 3);

    const result = collectingMetrics.collect();
    expect(result).toEqual({
      counters: { [counterResult1.metric.name]: counterResult1 },
      histograms: { [histResult1.metric.name]: histResult1 },
    });

    expect(collectingMetrics.collect()).toEqual(emptyMetrics);

    counter.inc(3);
    expect(collectingMetrics.collect()).toEqual({
      ...emptyMetrics,
      counters: {
        [counterResult1.metric.name]: {
          metric: counterResult1.metric,
          values: [{ countOrLabels: 3, count: undefined }],
        },
      },
    });
  });

  test('Reporter - POST success', async () => {
    (global as any).fetch = jest.fn(() => Promise.resolve({ ok: true }));

    reporter = new Reporter({
      logger,
      timer: 2000,
      endpoint: 'fakeEnd',
    });

    logger.log(options1);

    const counter = metrics.createCounter(counterOptions1);
    counter.inc(5);
    counter.inc(4);

    const firstResult = await reporter.collectReport_forTest({
      backReport: emptyBackReport,
      logger,
    });

    expect(firstResult).toEqual(emptyBackReport);

    const histogram = metrics.createHistogram(histogramOptions1);
    histogram.observe({ histLabel1: 14 }, 3);

    const secondResult = await reporter.collectReport_forTest({
      backReport: emptyBackReport,
      logger,
    });

    expect(secondResult).toEqual(emptyBackReport);

    expect((fetch as any).mock.calls).toMatchSnapshot();

    await new Promise((resolve) => setTimeout(() => resolve(), 2000));
    expect((fetch as any).mock.calls.length).toBeGreaterThanOrEqual(3);
  });

  test('Reporter - POST failure', async () => {
    (global as any).fetch = jest.fn(() => Promise.resolve({ ok: false }));

    reporter = new Reporter({
      logger,
      timer: 2000,
      endpoint: 'fakeEnd',
    });

    logger.log(options1);

    const counter = metrics.createCounter(counterOptions1);
    counter.inc(5);
    counter.inc(4);

    const firstBackReport = {
      logs: [options1],
      metrics: {
        counters: {
          [requestCounter.name]: {
            metric: requestCounter,
            values: [{ countOrLabels: 0 }],
          },
          [failuresCounter.name]: {
            metric: failuresCounter,
            values: [{ countOrLabels: 0 }],
          },
          [counterResult1.metric.name]: counterResult1,
        },
        histograms: {},
      },
    };

    const firstResult = await reporter.collectReport_forTest({
      backReport: emptyBackReport,
      logger,
    });

    expect(firstResult).toEqual(firstBackReport);

    const histogram = metrics.createHistogram(histogramOptions1);
    histogram.observe({ histLabel1: 14 }, 3);

    const secondBackReport = {
      logs: [],
      metrics: {
        counters: {
          [requestCounter.name]: {
            metric: requestCounter,
            values: [{ countOrLabels: undefined, count: undefined }],
          },
          [failuresCounter.name]: {
            metric: failuresCounter,
            values: [{ countOrLabels: undefined, count: undefined }],
          },
          [counterResult1.metric.name]: {
            metric: counterResult1.metric,
            values: [],
          },
        },
        histograms: { [histResult1.metric.name]: histResult1 },
      },
    };

    const secondResult = await reporter.collectReport_forTest({
      backReport: emptyBackReport,
      logger,
    });

    expect(secondResult).toEqual(secondBackReport);

    expect((fetch as any).mock.calls).toMatchSnapshot();

    await new Promise((resolve) => setTimeout(() => resolve(), 2000));
    expect((fetch as any).mock.calls.length).toBeGreaterThanOrEqual(3);
  });

  describe('report to Node', () => {
    test('Counter.inc called with same args', () => {
      const counter = metrics.createCounter(counterOptions1);
      const clientCounterInc = jest.spyOn(counter, 'inc');
      counter.inc(5);
      counter.inc(4);
      counter.inc({ label1: 7 });
      counter.inc({ label2: 12 }, 36);

      const testReport = JSON.stringify({
        logs: logger.collect(),
        metrics: collectingMetrics.collect(),
      });

      const nodeMonitor = NodeMonitor.create({
        service: 'test',
      });
      const nodeCounter = {
        inc: jest.fn(),
      };
      const createCounterSpy = jest.spyOn(metrics, 'createCounter');
      createCounterSpy.mockImplementationOnce(() => nodeCounter);
      nodeMonitor.report(JSON.parse(testReport));
      createCounterSpy.mockRestore();

      expect(clientCounterInc.mock.calls).toEqual(nodeCounter.inc.mock.calls);
    });

    test('Histogram.observe called with same args', () => {
      const histogram = metrics.createHistogram(histogramOptions1);
      const clientHistogramObserve = jest.spyOn(histogram, 'observe');
      histogram.observe({ histLabel1: 14 }, 3);
      histogram.observe(7);

      const testReport = JSON.stringify({
        logs: logger.collect(),
        metrics: collectingMetrics.collect(),
      });

      const nodeMonitor = NodeMonitor.create({
        service: 'test',
      });
      const nodeHistogram = {
        observe: jest.fn(),
      };
      const createHistogramSpy = jest.spyOn(metrics, 'createHistogram');
      createHistogramSpy.mockImplementationOnce(() => nodeHistogram);
      nodeMonitor.report(JSON.parse(testReport));
      createHistogramSpy.mockRestore();

      expect(clientHistogramObserve.mock.calls).toEqual(
        nodeHistogram.observe.mock.calls,
      );
    });

    test('logging', () => {
      logger.log(options1);
      logger.log(options2);

      const testReport = JSON.stringify({
        logs: logger.collect(),
        metrics: collectingMetrics.collect(),
      });
      const nodeMonitor = NodeMonitor.create({
        service: 'test',
      });
      const logSpy = jest.spyOn((nodeMonitor as any).logger, 'log');
      nodeMonitor.report(JSON.parse(testReport));

      expect([[options1], [options2]]).toEqual(logSpy.mock.calls);
    });
  });
});
