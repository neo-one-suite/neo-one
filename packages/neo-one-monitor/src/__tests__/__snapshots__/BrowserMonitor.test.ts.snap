// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserMonitor Reporter - POST failure 1`] = `
Array [
  Array [
    "fakeEnd",
    Object {
      "body": "{\\"logs\\":[{\\"name\\":\\"log1\\",\\"level\\":\\"info\\",\\"message\\":\\"test1\\",\\"labels\\":{\\"label1\\":10},\\"data\\":{\\"data1\\":\\"20\\"}}],\\"metrics\\":{\\"counters\\":{\\"http_report_metrics_requests_total\\":{\\"metric\\":{\\"name\\":\\"http_report_metrics_requests_total\\"},\\"values\\":[{\\"countOrLabels\\":0}]},\\"http_report_metrics_failures_total\\":{\\"metric\\":{\\"name\\":\\"http_report_metrics_failures_total\\"},\\"values\\":[{\\"countOrLabels\\":0}]},\\"counter1\\":{\\"metric\\":{\\"name\\":\\"counter1\\",\\"help\\":\\"helpCount\\",\\"labelNames\\":[\\"label1\\",\\"label2\\"]},\\"values\\":[{\\"countOrLabels\\":5},{\\"countOrLabels\\":4}]}},\\"histograms\\":{}}}",
      "method": "POST",
    },
  ],
  Array [
    "fakeEnd",
    Object {
      "body": "{\\"logs\\":[],\\"metrics\\":{\\"counters\\":{\\"http_report_metrics_requests_total\\":{\\"metric\\":{\\"name\\":\\"http_report_metrics_requests_total\\"},\\"values\\":[{}]},\\"http_report_metrics_failures_total\\":{\\"metric\\":{\\"name\\":\\"http_report_metrics_failures_total\\"},\\"values\\":[{}]},\\"counter1\\":{\\"metric\\":{\\"name\\":\\"counter1\\",\\"help\\":\\"helpCount\\",\\"labelNames\\":[\\"label1\\",\\"label2\\"]},\\"values\\":[]}},\\"histograms\\":{\\"histogram1\\":{\\"metric\\":{\\"name\\":\\"histogram1\\",\\"help\\":\\"helpHist\\",\\"labelNames\\":[\\"histLabel1\\",\\"histLabel2\\"]},\\"values\\":[{\\"countOrLabels\\":{\\"histLabel1\\":14},\\"count\\":3}]}}}}",
      "method": "POST",
    },
  ],
]
`;

exports[`BrowserMonitor Reporter - POST success 1`] = `
Array [
  Array [
    "fakeEnd",
    Object {
      "body": "{\\"logs\\":[{\\"name\\":\\"log1\\",\\"level\\":\\"info\\",\\"message\\":\\"test1\\",\\"labels\\":{\\"label1\\":10},\\"data\\":{\\"data1\\":\\"20\\"}}],\\"metrics\\":{\\"counters\\":{\\"http_report_metrics_requests_total\\":{\\"metric\\":{\\"name\\":\\"http_report_metrics_requests_total\\"},\\"values\\":[{\\"countOrLabels\\":0}]},\\"http_report_metrics_failures_total\\":{\\"metric\\":{\\"name\\":\\"http_report_metrics_failures_total\\"},\\"values\\":[{\\"countOrLabels\\":0}]},\\"counter1\\":{\\"metric\\":{\\"name\\":\\"counter1\\",\\"help\\":\\"helpCount\\",\\"labelNames\\":[\\"label1\\",\\"label2\\"]},\\"values\\":[{\\"countOrLabels\\":5},{\\"countOrLabels\\":4}]}},\\"histograms\\":{}}}",
      "method": "POST",
    },
  ],
  Array [
    "fakeEnd",
    Object {
      "body": "{\\"logs\\":[],\\"metrics\\":{\\"counters\\":{\\"http_report_metrics_requests_total\\":{\\"metric\\":{\\"name\\":\\"http_report_metrics_requests_total\\"},\\"values\\":[{}]},\\"http_report_metrics_failures_total\\":{\\"metric\\":{\\"name\\":\\"http_report_metrics_failures_total\\"},\\"values\\":[]},\\"counter1\\":{\\"metric\\":{\\"name\\":\\"counter1\\",\\"help\\":\\"helpCount\\",\\"labelNames\\":[\\"label1\\",\\"label2\\"]},\\"values\\":[]}},\\"histograms\\":{\\"histogram1\\":{\\"metric\\":{\\"name\\":\\"histogram1\\",\\"help\\":\\"helpHist\\",\\"labelNames\\":[\\"histLabel1\\",\\"histLabel2\\"]},\\"values\\":[{\\"countOrLabels\\":{\\"histLabel1\\":14},\\"count\\":3}]}}}}",
      "method": "POST",
    },
  ],
]
`;
